tests/pass/shims/x86/intrinsics-x86-bmi.rs

```diff
@@ -0,0 +1,224 @@
+// Ignore everything except x86 and x86_64
+// Any new targets that are added to CI should be ignored here.
+// (We cannot use `cfg`-based tricks here since the `target-feature` flags below only work on x86.)
+//@ignore-target-aarch64
+//@ignore-target-arm
+//@ignore-target-avr
+//@ignore-target-s390x
+//@ignore-target-thumbv7em
+//@ignore-target-wasm32
+//@compile-flags: -C target-feature=+bmi1,+bmi2
+
+#[cfg(target_arch = "x86")]
+use std::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
+
+fn main() {
+    // BMI1 and BMI2 are independent from each other, so both must be checked.
+    assert!(is_x86_feature_detected!("bmi1"));
+    assert!(is_x86_feature_detected!("bmi2"));
+
+    unsafe {
+        test_bmi_32();
+        test_bmi_64();
+    }
+}
+
+/// Test the 32-bit variants of the intrinsics.
+unsafe fn test_bmi_32() {
+    unsafe fn test_bextr_u32() {
+        let r = _bextr_u32(0b0101_0000u32, 4, 4);
+        assert_eq!(r, 0b0000_0101u32);
+
+        for i in 0..16 {
+            assert_eq!(_bextr_u32(u32::MAX, i, 4), 0b1111);
+            assert_eq!(_bextr_u32(u32::MAX, 4, i), (1 << i) - 1);
+        }
+
+        // Ensure that indices larger than the bit count are covered.
+        // It is important to go above 255 in order to verify the bit selection
+        // of the instruction.
+
+        for i in 0..=512 {
+            // The instruction only takes the lowest eight bits to generate the index, hence `i & 0xff`.
+            // If the index is out of bounds, the original input won't be changed, thus the `min(32)`.
+            assert_eq!(_bextr_u32(u32::MAX, 0, i).count_ones(), (i & 0xff).min(32));
+        }
+
+        for i in 0..=512 {
+            assert_eq!(_bextr_u32(u32::MAX, i, 0), 0);
+        }
+
+        for i in 0..=1024 {
+            let start = i & 0xff;
+            let len = (i >> 8) & 0xff;
+            let expected = {
+                let shifted = u32::MAX.checked_shr(start).unwrap_or(0);
+                if len >= 32 { shifted } else { shifted & ((1u32 << len) - 1) }
+            };
+            let actual = _bextr2_u32(u32::MAX, i);
+            assert_eq!(expected, actual);
+        }
```

Comment:
Sorry, but this doesn't make a ton of sense to me. This is a lot of loop iterations and still not covering some interesting cases, like when `i` is above 0xffff (and hence has some bits that are ignored entirely). Also this goes to show that the first loop, that goes until 512, is kind of pointless since `i` is already going to be masked before it even reaches the intrinsic.

I think it makes more sense to just test a few hand-picked cases that call the intrinsic directly with interesting values, and hard-code their results, with some comments explaining things.

Reply:
Done.

Reply:
> Also this goes to show that the first loop, that goes until 512, is kind of pointless since i is already going to be masked before it even reaches the intrinsic.

This part has not been resolved. It makes no sense for the loop to go beyond 255.

Reply:
The loop bounds are adjusted now.